cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ExpressionFactory.java

```diff
@@ -31,7 +31,15 @@ public Expression createExpression(String expressionString) {
         if (m.find()) {
             return new RegularExpression(Pattern.compile(m.group(1)), parameterTypeRegistry);
         }
-        return new CucumberExpression(expressionString, parameterTypeRegistry);
+        return createCucumberExpression(expressionString);
+    }
+
+    private Expression createCucumberExpression(String expressionString) {
+        try {
+            return new CucumberExpression(expressionString, parameterTypeRegistry);
+        } catch (UndefinedParameterTypeException e) {
```

Comment:
We're using exceptions for control flow. Might be worth extracting some of the code in the `CucumberExpression` to the factory instead of throwing the exception and then catching it.

Reply:
Yeah I had the same thought.